### DISCUSSION TITLE
Poetry VS Pixi when building Python CLI apps

### DISCUSSION THREAD

**Discussion by T145** [prefix-dev/pixi#2392] [Q&A] [answered]
When using Poetry, I can have something like this in my `pyproject.toml`:
```
[tool.poetry.scripts]
myproj = "myproj.main:app"
```
Where `app` is a CLI function in `main`, and after installing be able to run `myproj`.
Using Pixi's task, to my knowledge this is the equivalent:
```
[tool.pixi.tasks]
myproj = "python src/myproj/main.py"
```
And it still requires `pixi run myproj`. Is there a way to accomplish what Poetry does with Pixi?

**Comment by synapticarbors** [comment]
My understanding is that `[tool.poetry.scripts]` is a python build tool way of creating an entry point. You can do the same using other build backends (e.g [hatch](https://hatch.pypa.io/latest/config/metadata/#cli). See also the [entry point spec](https://packaging.python.org/en/latest/specifications/entry-points/#entry-points-specification)).

It seems it would be best practice to use one of the build backends to install your local project, here contained in `src`, into a pixi env, and then it would just be available as `myproj` on the command line once you enter `pixi shell` or in a task. This can either be done with a `pypi-dependencies` that specifies the local project (see: https://github.com/prefix-dev/pixi/blob/main/examples/pypi-source-deps/pixi.toml#L29-L30), or via a task that uses pip or uv to install it, like in https://github.com/prefix-dev/pixi/blob/main/examples/docker/pyproject.toml#L46-L48

**Reply by T145** [reply]
Ok so the first example is interesting! I have a Python project with only `pyproject.toml` containing the project settings. When creating a Python CLI project, do I then need to create an initial Pixi project with `pixi init`, then do `pixi init myproj --pyproject`?

**Comment by jesserobertson** [comment, marked as answer]
FWIW I found this a bit non-obvious, although a combo of stackoverflow and hatchling docs got me there in the end. 

In case it helps someone looking for the magic commands to include, here's what worked for me. If I have a hello world CLI in `src/PACKAGE/cli.py`:

```python
# file: src/PACKAGE/cli.py
import typer

cli = typer.Typer()

@cli.command()
def hello(name: str):
    typer.echo(f"Hello {name}")

if __name__ == "__main__":
    cli()
```

...then this is what I put in the pyproject.toml:

```toml
# ...
[build-system]
build-backend = "hatchling.build"
requires = ["hatchling"]

[project.scripts]
pkg_cli = "PACKAGE.cli:cli"
# ...
```

...and I can just run the command in the pixi shell (or when it gets installed etc):

```bash
$ pixi shell 
(env) $ pkg_cli "world!"
Hello world!
```
